### ISSUE TITLE
ModDB download link not found

### ISSUE THREAD

**Issue by allumik** [Mord3rca/gamma-launcher#167]
Master branch on NixOS

Looking at GAMMA discord then it might be related to the ModDB changing the security policy?

Is it possible to use the 10-30 downloads archive that Grok has shared instead of downloading from moddb?

```
gamma-launcher full-install --anomaly Anomaly --gamma Gamma
[+] Updating G.A.M.M.A. definition
[+] Reading mod definition from Gamma/.Grok's Modpack Installer/G.A.M.M.A/modpack_data ...
    Fetching remote Grokitach from gamma_large_files_v2.git...
[+] Installing Git Mod: https://github.com/Grokitach/gamma_large_files_v2
  --> Installing 273- Longreed's fixed artefacts spawns under maps - Longreed
  --> Installing 272- Grulag's Dead Bushes - Grulag
  --> Installing G.A.M.M.A. Large Files
  --> Installing Boomsticks and Sharpsticks 2022 Guns
  --> Installing Burn's Optimised World Models
  --> Installing Retrogue's Additional Weapons
    Fetching remote Grokitach from teivaz_anomaly_gunslinger.git...
[+] Installing Git Mod: https://github.com/Grokitach/teivaz_anomaly_gunslinger
  --> Installing 312- Gunslinger Guns for Anomaly - Teivazcz & Gunslinger Team
[+] Installing mod: Attachment Alignment Fixes
Extracting Gamma/downloads/Tiskar-s-Burn-s-Attachment-Alignment-Fixes-master.zip
    Installing Tiskar-s-Burn-s-Attachment-Alignment-Fixes-master -> Gamma/mods/368- Attachment Alignment Fixes - Tiskar & Burn
Traceback (most recent call last):
  File "/nix/store/xwsa42x24c4022rvkwh07m5fmqrp0d65-python3.12-gamma-launcher-86c2d1b22df8da901b8202b01e31f74ac78cef95/bin/.gamma-launcher-wrapped", line 9, in <module>
    sys.exit(main())
             ^^^^^^
  File "/nix/store/pgfpk7pzcz52h0dxipvcz5gvajlh07hh-python3-3.12.4-env/lib/python3.12/site-packages/launcher/cli.py", line 113, in main
    args.cobject().run(args)
  File "/nix/store/pgfpk7pzcz52h0dxipvcz5gvajlh07hh-python3-3.12.4-env/lib/python3.12/site-packages/launcher/commands/install.py", line 305, in run
    self._install_mods()
  File "/nix/store/pgfpk7pzcz52h0dxipvcz5gvajlh07hh-python3-3.12.4-env/lib/python3.12/site-packages/launcher/commands/install.py", line 253, in _install_mods
    mod.download(self._dl_dir, use_cached=True)
  File "/nix/store/pgfpk7pzcz52h0dxipvcz5gvajlh07hh-python3-3.12.4-env/lib/python3.12/site-packages/launcher/mods/downloader/moddb.py", line 97, in download
    self._url = self._get_download_url(self._url)
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/nix/store/pgfpk7pzcz52h0dxipvcz5gvajlh07hh-python3-3.12.4-env/lib/python3.12/site-packages/launcher/mods/downloader/moddb.py", line 49, in _get_download_url
    raise ModDBDownloadError(f"Download link not found when requesting {url}")
launcher.exceptions.ModDBDownloadError: Download link not found when requesting https://www.moddb.com/addons/start/208873
```

**Comment by Vponed** [comment]
I have the same problem, I don't know what to do

**Comment by Mord3rca** [comment]
Hello,

Can you open the link in a browser when this happen ?

**Comment by Mord3rca** [comment]
I'm gonna need ModDB node IP addr to test that since I don't have any issue on my side and it may be related to ModDB mirror like in #42 

Regarding cache directory:
You won't be able to do "offline" install (since packages names are fetch on ModDB)
Usage looks fine, make sure Gamma/downloads is a symlink to your cache folder.
Which mods are being refetch ? Maybe they are updated or something, ...

**Comment by allumik** [comment]
Thanks for the very quick reply and explaining why the --cache-directory option still needs to contact modDB!

The default for me seems to be the NL mirror 104.26.13.171; `dig AAAA www.moddb.com`:
```
www.moddb.com.          283     IN      AAAA    2606:4700:20::681a:dab
www.moddb.com.          283     IN      AAAA    2606:4700:20::681a:cab
www.moddb.com.          283     IN      AAAA    2606:4700:20::ac43:464e
```

I tried VPN and #42 solution but no luck. The node seems to be the same as you suggested. Specifying the host does give me 443 instead of 403, both in VPN and local (after 2x previous calls of `gamma-launcher`, hope it did not block me yet :grimacing:).

Regarding the gamma-launcher installation - I do have to note that I modified the `pyproject.toml` when writing definition files for the package in nix, renaming `bs4` to its alias `beautifulsoup4` ("bs4" is not in the nix packages, "beautifulsoup4" is) but I doubt it makes any difference:

```
self: super: {
  "gamma-launcher" = super.buildPythonPackage rec {
  ...
    preBuild = ''sed -i "s/bs4/beautifulsoup4/g" pyproject.toml'';
  ...
```

**Comment by Angelthree95** [comment]
I have the same problem, but trying to do a fresh install. 
```
[+] Installing base Anomaly 1.5.3
Traceback (most recent call last):
  File "gamma-launcher.py", line 8, in <module>
  File "launcher/cli.py", line 113, in main
  File "launcher/commands/install.py", line 293, in run
  File "launcher/commands/install.py", line 64, in run
  File "launcher/mods/downloader/moddb.py", line 97, in download
  File "launcher/mods/downloader/moddb.py", line 49, in _get_download_url
launcher.exceptions.ModDBDownloadError: Download link not found when requesting https://www.moddb.com/downloads/start/277404
[PYI-124974:ERROR] Failed to execute script 'gamma-launcher' due to unhandled exception!
```
I have Anomaly 1.5.3 in the cache folder already, so idk why would the launcher re-download.

wget yields a 403:
```
wget https://www.moddb.com/downloads/start/277404
--2024-11-13 00:08:19--  https://www.moddb.com/downloads/start/277404
Translacja www.moddb.com (www.moddb.com)... 104.26.13.171, 104.26.12.171, 172.67.70.78, ...
Łączenie się z www.moddb.com (www.moddb.com)|104.26.13.171|:443... połączono.
Żądanie HTTP wysłano, oczekiwanie na odpowiedź... 403 Forbidden
2024-11-13 00:08:19 BŁĄD 403: Forbidden.
```
I'm working with the latest release (v2.0)

**Comment by Mord3rca** [comment]
There is a no-flood indeed ... some request are blocked with a "not too fast" page even if this is the 2nd request ...
It seems to be on all nodes.

Gonna need to patch that

**Comment by Mord3rca** [comment]
@Angelthree95 Even if you have the file in cache, a ModDB request is done to get the filename & MD5 hash of it. This is why you have a crash even if you already have the file.

**Comment by Mord3rca** [comment]
Good news everyone,

Can you try PR #170 ? It should fix the Cloudflare IUAM issue and fix ModDB DL.

Note: 2nd request will take longer (5 seconds) but after this, all clear.

**Comment by allumik** [comment]
It works for me :partying_face: 

Good timing, as the *other* game that got available to download today probably needs a few months to stabilise...

I did try out adding [tenacity decorators](https://tenacity.readthedocs.io/en/latest/) to the fetchers, which ofcourse did not work, and realised that its probably requiring authentification. I did not know of the cloudflare workaround and ran out of time to test further by then.

Thanks for the fix and also thanks for the beautifulsoup4 naming change.

Closing as it seems to be resolved.
